src/svd/parsing.py

```diff
@@ -4,34 +4,128 @@
         raise FileNotFoundError(f"No such file: {svd_file.absolute()}")
 
-    device = ET.parse(svd_file).getroot()
-    peripheral = Peripheral(device, peripheral_name)
+    try:
+        # Note: remove comments as otherwise these are present as nodes in the returned XML tree
+        xml_parser = objectify.makeparser(remove_comments=True)
+        class_lookup = _TwoLevelTagLookup(bindings.BINDINGS)
+        xml_parser.set_element_class_lookup(class_lookup)
+
+        with open(svd_file, "r") as f:
+            xml_device = objectify.parse(f, parser=xml_parser)
+
+        device = Device(xml_device.getroot(), options=options)
+
+    except Exception as e:
+        raise SvdParseError(f"Error parsing SVD file {svd_file}") from e
+
+    return device
+
+
+class _TwoLevelTagLookup(ET.ElementNamespaceClassLookup):
+    """
+    XML element class lookup that uses two levels of tag names to map an XML element to a Python
+    class. This two-level scheme is used to slightly optimize the time spent by the parser looking
+    up looking up the class for an element (which is a sizeable portion of the time spent parsing).
+
+    Element classes that can be uniquely identified by tag only are stored in the first level.
+    This level uses the lxml ElementNamespaceClassLookup which is faster than the second level.
+    The remaining element classes are assumed to be uniquely identified by a combination of
+    the parent tag and the tag itself, and are stored in the second level.
+    The second level uses the lxml PythonElementClassLookup which is slower.
+    """
+
+    def __init__(self, element_classes: List[Type[objectify.ObjectifiedElement]]):
+        """
+        :param element_classes: lxml element classes to add to the lookup table.
+        """
+        super().__init__()
+
+        tag_classes: Dict[str, Set[type]] = defaultdict(set)
+        two_tag_classes: Dict[Tuple[Optional[str], str], Set[type]] = defaultdict(set)
+
+        for element_class in element_classes:
+            tag = element_class.TAG
+            tag_classes[tag].add(element_class)
+            for prop in bindings.get_binding_elem_props(element_class).values():
+                tag_classes[prop.name].add(prop.klass)
+                two_tag_classes[(tag, prop.name)].add(prop.klass)
+
+        one_tag: Set[str] = set()
+        namespace = self.get_namespace(None)  # note: None is the empty namespace
+
+        for tag, classes in tag_classes.items():
+            if len(classes) == 1:
+                # Add the class to the namespace
+                element_class = classes.pop()
+                # note: namespace is a decorator, so the syntax here is a little odd
+                namespace(tag)(element_class)
+                one_tag.add(tag)
+
+        two_tag_lookup: Dict[Tuple[Optional[str], str], type] = {}
+
+        for (tag, field_name), classes in two_tag_classes.items():
+            if field_name in one_tag:
+                continue
+
+            assert (
```

Comment:
Maybe a personal preference, but `assert` feels out of place when not in tests and has a weird syntax that makes it easy to forget adding a useful string (not the case here, but promoting their use opens that box). Wouldn't this really be an IndexError or ValueError, since a length other than 1 is invalid?